R/get_nldi.R

```diff
@@ -115,28 +115,57 @@ navigate_nldi <- function(nldi_feature, mode = "upstreamMain",
 #' @details Only resolves to the nearest NHDPlus catchment divide. See:
 #' https://waterdata.usgs.gov/blog/nldi-intro/ for more info on the nldi.
 #' @inheritParams navigate_nldi
+#' @param simplify logical should response geometry be simplified for
+#' visualization and performance?
+#' @param split logical should response resolve precisely to the location
+#' of the `nldi_feature`? Setting `TRUE` calls an aditional service and
+#' will be slower and less robust.
 #' @return sf data.frame with result basin boundary
 #' @export
+#' @importFrom sf read_sf
 #' @examples
 #' \donttest{
 #' library(sf)
 #' library(dplyr)
 #'
 #' nldi_nwis <- list(featureSource = "nwissite", featureID = "USGS-05428500")
 #'
+#' site <- get_nldi_feature(nldi_nwis)
+#'
 #' basin <- get_nldi_basin(nldi_feature = nldi_nwis)
 #'
-#' basin %>%
-#'  st_geometry() %>%
-#'  plot()
+#' plot(st_geometry(basin))
 #'
 #' basin
+#'
+#' basin2 <- get_nldi_basin(nldi_feature = nldi_nwis,
+#'                          simplify = FALSE, split = TRUE)
+#'
+#' length(st_coordinates(basin))
+#' length(st_coordinates(basin2))
+#'
+#' plot(st_geometry(st_buffer(site, units::set_units(3000, "m"))), border = NA)
+#'
+#' plot(st_geometry(site), add = TRUE)
+#' plot(st_geometry(basin2), add = TRUE)
+#'
+#' plot(st_geometry(basin), border = "red", add = TRUE)
+#'
 #' }
-get_nldi_basin <- function(nldi_feature) {
+get_nldi_basin <- function(nldi_feature, simplify = TRUE, split = FALSE) {
 
-  nldi_feature <- check_nldi_feature(nldi_feature)
+  nldi_feature <- check_nldi_feature(nldi_feature, convert = FALSE)
+
+  sim <- "true"
+  spl <- "false"
+
+  if(!simplify) sim <- "false"
+  if(split) spl <- "true"
 
-  dataRetrieval::findNLDI(origin = nldi_feature, find = "basin")$basin
+  read_sf(query_nldi(paste0(nldi_feature[["featureSource"]], "/",
+                            nldi_feature[["featureID"]], "/basin?simplify=",
+                            sim, "&splitCatchment=", spl),
+                     parse_json = FALSE))
```

Comment:
Using query nldi here for basins so we have access to split and simplify. Might plow this into dataRetrieval too, but the more sophisticated client here kind of makes sense.